server/src/main/java/app/challenge/ChallengeService.java

```diff
@@ -0,0 +1,123 @@
+package app.challenge;
+
+import app.user.User;
+import app.user.UserRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class ChallengeService {
+    private final ChallengeRepository challengeRepository;
+    private final ChallengeParticipantsRepository challengeParticipantsRepository;
+    private final UserRepository userRepository;
+
+    public void addChallenge(Integer userId, ChallengeRequest challengeRequest) {
+        Challenge challenge = Challenge.builder()
+            .title(challengeRequest.title())
+            .description(challengeRequest.description())
+            .goal(challengeRequest.goal())
+            .unit(challengeRequest.unit())
+            .createdAt(ZonedDateTime.now())
+            .build();
+
+        challengeRepository.save(challenge);
+
+        List<User> participants = userRepository.findAllById(challengeRequest.participants());
+        Map<Integer, User> userMap = participants.stream()
+            .collect(Collectors.toMap(User::getId, user -> user));
+
+        addParticipantsToChallenge(userMap, challenge, challengeRequest.participants());
+
+        User creator = userRepository.findById(userId)
+            .orElseThrow(() -> new UsernameNotFoundException("User not found"));
```

Comment:
first get users and maybe throw error then do challengeRepository.save(challenge) 

why ? because otherwise we can create challenge  and stop it in a middle if partisipant was not found.  it is important to always think about database in this endpoints as database state should always be saved at the last step if possible

Reply:
in ideal world we would have **transactions** for database, so you can save any time and commit transaction at the end of endpoint when there is no errors, if there is an error all changes will be reverted. but now we dont have time to create a transaction and to set up them.